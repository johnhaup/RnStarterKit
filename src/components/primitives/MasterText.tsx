import { compact } from 'lodash';
import get from 'lodash/get';
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import React from 'react';
import { StyleProp, StyleSheet, TextProps, TextStyle } from 'react-native';
import Animated from 'react-native-reanimated';
import { ColorPalette } from '../../styles';

interface Props extends TextProps {
  children: string | undefined | null;
  style?: StyleProp<TextStyle>;
  xs?: boolean;
  sm?: boolean;
  lg?: boolean;
  xl?: boolean;
  red?: boolean;
  gray?: boolean;
  white?: boolean;
  bold?: boolean;
  medBold?: boolean;
  color?: string | Animated.Node<number>;
}

const styleSheetProps = [
  'xs',
  'sm',
  'lg',
  'xl',
  'gray',
  'white',
  'red',
  'bold',
  'medBold',
];

export const MasterText = ({ children, style, color, ...rest }: Props) => {
  // If value sent is an optional from an object and is not defined, render null
  // instead of an empty component
  if (!children) {
    return null;
  }

  const flaggedStyles = pick(rest, styleSheetProps);
  const dynamicStyleKeys = compact(Object.keys(flaggedStyles));
  const dynamicStyleSheets: TextStyle[] = compact(
    dynamicStyleKeys.map((key) => get(styles, key, null)),
  );
  const textStyles = [styles.base, { color }, ...dynamicStyleSheets, style];

  const textComponentProps = omit(rest, styleSheetProps);

  return (
    <Animated.Text style={textStyles} {...textComponentProps}>
      {children}
    </Animated.Text>
  );
};

const styles = StyleSheet.create({
  base: {
    // fontFamily: PUT YOUR CUSTOM BASE FONT HERE,
    fontSize: 15,
    lineHeight: 20,
    color: ColorPalette.black,
  },
  xs: {
    fontSize: 11,
    lineHeight: 15,
  },
  sm: {
    fontSize: 13,
    lineHeight: 18,
  },
  lg: {
    fontSize: 17,
    lineHeight: 23,
  },
  xl: {
    fontSize: 24,
    lineHeight: 30,
    fontWeight: '800',
  },
  gray: {
    color: ColorPalette.gray,
  },
  red: {
    color: ColorPalette.red,
  },
  white: {
    color: ColorPalette.white,
  },
  bold: {
    fontWeight: '800',
  },
  medBold: {
    fontWeight: '500',
  },
});
