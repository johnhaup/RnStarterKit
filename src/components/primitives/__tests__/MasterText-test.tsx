/**
 * @format
 */

import React from 'react';
import 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { MasterText } from '../MasterText';

it('renders correctly', () => {
  const snap = renderer.create(
    <MasterText xs sm lg xl red gray white bold medBold>
      Hello
    </MasterText>,
  );
  expect(snap).toMatchSnapshot();
});

it('renders correctly with props', () => {
  const snap = renderer.create(<MasterText>Hello</MasterText>);
  expect(snap).toMatchSnapshot();
});

it('return null if children are null', () => {
  const componentWithoutChildren = renderer
    .create(<MasterText>{null}</MasterText>)
    .getInstance();
  expect(componentWithoutChildren).toEqual(null);
});
