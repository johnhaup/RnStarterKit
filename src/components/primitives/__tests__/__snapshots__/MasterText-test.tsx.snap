// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#2d3436",
        "fontSize": 15,
        "lineHeight": 20,
      },
      Object {
        "color": undefined,
      },
      Object {
        "fontSize": 11,
        "lineHeight": 15,
      },
      Object {
        "fontSize": 13,
        "lineHeight": 18,
      },
      Object {
        "fontSize": 17,
        "lineHeight": 23,
      },
      Object {
        "fontSize": 24,
        "fontWeight": "800",
        "lineHeight": 30,
      },
      Object {
        "color": "#b2bec3",
      },
      Object {
        "color": "#ffffff",
      },
      Object {
        "color": "#d63031",
      },
      Object {
        "fontWeight": "800",
      },
      Object {
        "fontWeight": "500",
      },
      undefined,
    ]
  }
>
  Hello
</Text>
`;

exports[`renders correctly with props 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#2d3436",
        "fontSize": 15,
        "lineHeight": 20,
      },
      Object {
        "color": undefined,
      },
      undefined,
    ]
  }
>
  Hello
</Text>
`;
